tv_apps/tv_shared/components/smart-dialog/script.js

```diff
@@ -0,0 +1,184 @@
+'use strict';
+
+window.SmartDialog = (function(win) {
+
+  // Extend from the HTMLElement prototype
+  var proto = Object.create(HTMLElement.prototype);
+
+  /**
+   * gernerate bezier curve with corresponding parameters
+   * @param {[Object]} bezier curve parameters (p1, p2, p3, p4)
+   */
+  win.GenerateBezierCurve = function(bezierParam) {
+    var p1 = bezierParam.p1 || 0;
+    var p2 = bezierParam.p2 || 0;
+    var p3 = bezierParam.p3 || 0;
+    var p4 = bezierParam.p4 || 0;
+    var a1 = 3.0 * p1;
+    var b1 = 3.0 * (p3 - p1) - a1;
+    var c1 = 1.0 - a1 - b1;
+    var a2 = 3.0 * p2;
+    var b2 = 3.0 * (p4 - p2) - a2;
+    var c2 = 1.0 - a2 - b2;
+
+    var curveX = function(t) {
+      return ((c1 * t + b1) * t + a1) * t;
+    };
+
+    var curveY = function(t) {
+      return ((c2 * t + b2) * t + a2) * t;
+    };
+
+    var calculateCurve = function(x, epsilon) {
+      var t0;
+      var t1;
+      var t2;
+      var x2;
+       
+      t0 = 0.0;
+      t1 = 1.0;
+      t2 = x;
+       
+      if (t2 < t0) {
+        return t0;
+      }
+      if (t2 > t1) {
+        return t1;
+      }
+       
+      while (t0 < t1) {
+        x2 = curveX(t2);
+        if (Math.abs(x2 - x) < epsilon) {
+          return t2;
+        }
+        if (x > x2) {
+          t0 = t2;
+        } else {
+          t1 = t2;
+        }
+        t2 = (t1 - t0) * 0.5 + t0;
+      }
+       
+      return t2;
+    };
+     
+    return function(x) {
+      return curveY(calculateCurve(x, 0.001));
+    };
+  };
+
+  proto.circleAnimation = function(element, duration, bezierParam, callback) {
+
+    var winHalfWidth = window.innerWidth / 2;
+    var winHeight = window.innerHeight;
+    var finalSize = Math.sqrt((winHalfWidth * winHalfWidth +
+                               winHeight * winHeight));
+
+    // create svg element
+    var svgEle = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
+    var existingSvg = document.getElementById((element.id || 'smart-dialog') +
+                                               '-svg');
+    if (existingSvg) {
+      document.body.removeChild(existingSvg);
+    }
+    svgEle.setAttribute('id', element.id + '-svg');
+    svgEle.setAttribute('width', 0);
+    svgEle.setAttribute('height', 0);
+
+    // create clipPath element with clip-path id
+    var clipPathEle = document.createElementNS('http://www.w3.org/2000/svg',
+                                               'clipPath');
+    clipPathEle.setAttribute('id', element.id + '-clipCircle');
+
+    // create circle element as the shape of element border
+    var circleEle = document.createElementNS('http://www.w3.org/2000/svg',
+                                             'circle');
+    circleEle.setAttribute('cx', winHalfWidth);
+    circleEle.setAttribute('cy', winHeight);
+
+    // set element clip-path style
+    element.style.clipPath = 'url(\"#' + element.id + '-clipCircle\")';
```

Comment:
Since the element.id is MUST, we should return the function at the begin while no element.id.